### PR TITLE
Place and get structure endpoint

### PR DESCRIPTION
**Pull request by Niels-NTG**
Implements interfaces for the [`POST /structure`](https://github.com/Niels-NTG/gdmc_http_interface/blob/master/docs/Endpoints.md#place-nbt-structure-file-post-structure) and [`GET /structure`](https://github.com/Niels-NTG/gdmc_http_interface/blob/master/docs/Endpoints.md#create-nbt-structure-file-get-structure) endpoints present in GDMC-HTTP.

This doesn't currently include any integration with the `Editor` class or an util method that help with loading and saving `*.nbt` files. I would want to include these as well but don't know where best to put these.

### CODE REVIEW

**Review by avdstaaij — CHANGES_REQUESTED** [review]
The `interface.py` wrappers look good. I do however have a bunch of comments, mostly about the docstrings. In particular, I think we should be precise about the difference between a "file" and "the contents of a file", and the difference between "an nbt file" and "a structure file" (this is especially important because _schematic_ files are also nbt files).

There should also be a higher-level interface for structure files in the `Editor` class, making use of its transform. But that can also be added after this PR.

About the `nbt_tools.py` additions: I am generally against adding functions that are not used anywhere, but some of these are indeed quite useful.

**Comment by avdstaaij** [comment]
By the way, a question: do structures in the GDMC-HTTP mod have a maximum size like they do in-game? (48x48x48 in Java edition according to the wiki.) If so, what happens when you request a structure larger than that?

**Comment by Niels-NTG** [comment]
> By the way, a question: do structures in the GDMC-HTTP mod have a maximum size like they do in-game? (48x48x48 in Java edition according to the wiki.) If so, what happens when you request a structure larger than that?

During the development of the `POST /structure` and `GET /structure` endpoint I didn't encounter any of those limits. I managed to "copy-paste" entire villages including all inhabitants, much larger than 48x48x48 blocks, without issue. As far as I know the 48x48x48 limit only applies to structures that can be saved using the in-game [Structure Block](https://minecraft.fandom.com/wiki/Structure_Block). This limitation can be overcome by creating structure files via other methods, such as the `GET /structure` endpoint or mods like WorldEdit.

**Comment by avdstaaij** [comment]
That's good news!

File: gdpc/interface.py [hunk 1]
```diff
@@ -230,6 +230,78 @@ def getChunks(position: Vec2iLike, size: Optional[Vec2iLike] = None, dimension:
...
+    See the GDMC HTTP API documentation for more information about these parameters:
+    https://github.com/Niels-NTG/gdmc_http_interface/blob/master/docs/Endpoints.md#place-nbt-structure-file-post-structure
+    """
+    url = f"{host}/structure"
+    x, y, z = position
+    mirrorArg = None
+    if mirror[0]:
+        mirrorArg = 'x'
+    elif mirror[1]:
+        mirrorArg = 'z'
```
Comment (avdstaaij) [inline gdpc/interface.py:249]: This looks like it does not support simultaneous x and z mirroring. Does GDMC-HTTP support simultaneous mirroring? If not, it should be emulated here by converting it to the equivalent rotation.

File: gdpc/nbt_tools.py [hunk 2]
```diff
@@ -31,3 +31,30 @@ def nbtToSnbt(tag: nbt.TAG) -> str:
...
+    filePath: Union[Path, str]
+):
+    """Create NBT object from stored NBT file."""
+    if isinstance(filePath, str):
+        filePath = Path(filePath)
+    fileObject = open(filePath, 'rb')
+    return nbt.NBTFile(fileobj=fileObject)
+
+
+def saveNbtTFile(
```
Comment (avdstaaij) [inline gdpc/nbt_tools.py:46]: I think you missed a letter here ;)

**Review by avdstaaij — CHANGES_REQUESTED** [review]

File: gdpc/interface.py [hunk 3]
```diff
@@ -230,6 +230,81 @@ def getChunks(position: Vec2iLike, size: Optional[Vec2iLike] = None, dimension:
...
+    See the GDMC HTTP API documentation for more information about these parameters:
+    https://github.com/Niels-NTG/gdmc_http_interface/blob/master/docs/Endpoints.md#place-nbt-structure-file-post-structure
+    """
+    url = f"{host}/structure"
+    x, y, z = position
+    rotate = (rotate % 4) if rotate else None
+    mirrorArg = None
+    if mirror[0] and mirror[1]:
+        mirrorArg = 'z'
+        rotate = 1
```
Comment (avdstaaij) [inline gdpc/interface.py:249]: I believe this is still not correct: mirroring in one axis changes the structure's "orientation", while mirroring in two axes does not. This currently also overwrites the original `rotate` value. I'm pretty sure this should be:
```python
if mirror[0] and mirror[1]:
    rotate = (rotate + 2) % 4
```

**Review by avdstaaij — CHANGES_REQUESTED** [review]

**Comment by avdstaaij** [comment]
One more question: in what order does the GDMC-HTTP endpoint do mirroring and rotating? Mirror -> rotate or rotate -> mirror? (These are not equivalent!)
Every system in GDPC that has rotation and mirroring does mirroring first and rotation second, so if GDMC-HTTP works the other way around, we need to do some thinking to correct for that here.

**Comment by Niels-NTG** [comment]
> One more question: in what order does the GDMC-HTTP endpoint do mirroring and rotating? Mirror -> rotate or rotate -> mirror? (These are not equivalent!) Every system in GDPC that has rotation and mirroring does mirroring first and rotation second, so if GDMC-HTTP works the other way around, we need to do some thinking to correct for that here.

Looking at the `transform` method in `StructureTemplate.java` reveals that it first applies the mirror, than the rotation.

File: gdpc/interface.py (hunk 3 above)
Comment (Niels-NTG) [reply gdpc/interface.py]: :facepalm: yes, of course…

**Review by avdstaaij — APPROVED** [review]

**Pull request merged**